pkg/containers/containers.go

```diff
@@ -182,62 +202,112 @@ func (c *Containers) populate() error {
 // saving container information in Containers CgroupInfo map.
 // NOTE: ALL given cgroup dir paths are stored in CgroupInfo map.
 func (c *Containers) CgroupUpdate(cgroupId uint64, path string, ctime time.Time) (CgroupInfo, error) {
-	containerId, runtime := getContainerIdFromCgroup(path)
+	containerId, containerRuntime := getContainerIdFromCgroup(path)
+	container := cruntime.ContainerMetadata{
+		ContainerId: containerId,
+	}
 
 	info := CgroupInfo{
-		Path:        path,
-		ContainerId: containerId,
-		Runtime:     runtime,
-		Ctime:       ctime,
+		Path:      path,
+		Container: container,
+		Runtime:   containerRuntime,
+		Ctime:     ctime,
 	}
 
 	c.mtx.Lock()
 	c.cgroups[uint32(cgroupId)] = info
 	c.mtx.Unlock()
 
+	err := c.enrichCgroupInfo(cgroupId)
+
+	//don't fail on enrichment but log to stderr
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to enrich container data: %v", err)
+	}
+
 	return info, nil
 }
 
+// enrichCgroupInfo checks for a given cgroupId if it is relevant to some running container
+// it then calls the runtime info service to gather additional data from the container's runtime
+func (c *Containers) enrichCgroupInfo(cgroupId uint64) error {
+	c.mtx.RLock()
+	info, ok := c.cgroups[uint32(cgroupId)]
+	c.mtx.RUnlock()
+
+	//if there is no cgroup anymore for some reason, return early
+	if !ok {
+		return fmt.Errorf("no cgroup to enrich")
+	}
+
+	containerId := info.Container.ContainerId
+	runtime := info.Runtime
+
+	if containerId != "" && runtime != cruntime.Unknown {
+		//There might be a performance overhead with the cancel
+		//But, I think it will be negligable since this code path shouldn't be reached too frequently
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```

Comment:
Does this mean that the pipeline is stuck until we get the response (possibly for a few seconds)?
Won't this cause us loosing many events?
Maybe we should consider processing other events somehow while we're waiting here

Reply:
I think this code path is in a separate goroutine than event processing, so no? 
Maybe we should make sure enrichment always gets it's own goroutine and then we'll be fine, but I'm not sure it's necessary.

Reply:
I actually don't think it's in a separate goroutine, so this is why I'm asking.
If adding a new go routine, you should note that we might have event ordering issues.
In addition, if another event comes from the same container, you need to make sure that you don't query the runtime again.
So a proper solution might require:
1. Marking that there's an ongoing query for a given container (can be saved as a bool in the cgroup info)
2. Queue events from that container while there's an ongoing query

Reply:
> 1. Marking that there's an ongoing query for a given container (can be saved as a bool in the cgroup info)

In addition to that, if we already tried to query a given container id and failed in the past (for whatever reason that the runtime isn't aware of the container, or we might have a wrong container id parsed from the cgroup name), I think we should not try to query again. We can keep this information in the cgroupinfo as well

Reply:
I think the solution is refactoring this to be a part of the event pipeline, WDYT?

Reply:
There might be other ways to solve this, but this is a possible solution.
We can have another "enrichment" pipeline stage.

Reply:
Nice catch @yanivagman. So, the idea here would be to cache ALL events from a specific container while it is being enriched ? And then release them once the enrichment succeeds or fails ? If that is the case, I would place this as a last step in the pipeline (after sorting and caching). 

Be aware of the following as well: When we release all the events... we might have held up to 10 seconds of events (timeout for the enrichment to fail ?). This means that we could be releasing a big amount of data (max throughput * 10, could be something like 40MB in worst cases) and this could also cause latencies big latency in the pipeline as well.

Reply:
There are probably other ways to do it, but i'm going for a strategy similar to what @rafaeldtinoco described currently as a stage in the pipeline.

Reply:
@NDStrahilevitz maybe we should reduce the timeout to the absolute needed time. And mark the timedout ones not to re-try them over and over like @yanivagman suggested.

Reply:
I'm wondering if we should let the timeout(s) be configurable under the `crs` (that will be the new flag name) flag.
In addition, I am currently marking falied tries in the pipeline stage.

Currently, there's no good stage in the pipeline to add this logic, since I need the CgroupMkdir logic to run first, which is in the `processEvents` part. So I will need a PR to split `processEvents` from the pipeline sink stage (they are currently comined), and then I can put this enrich stage in between. In general, I think we could benefit from refactoring the pipeline to some degree, see #1684 that I opened.

Reply:
Updating that I have solved this through #1689. I need it merged so I can rebase this PR on top of it.

Reply:
PR #1689 has been merged. Please rebase this one and highlight the specific changes about:

1. enqueueing events when a specific event+containerid is being enriched
2. marking enrichment has failed and wont be tried again

so we can move on with this PR as well. Thanks!

PS:

Be aware that the enrichment process will cause latencies in the pipeline (unfortunately) as we can't queue events that need enrichment AND still process the subsequent ones, that don't need, because of chronological/ordering needs. 

IMO, we should "enforce" caching in the pipeline when using container enrichment (this would allow enrichment to take sometime while reducing the pressure being put by the perfbuffer in the origin).